src/connectors/impls/clickhouse/conversion.rs

```diff
@@ -0,0 +1,442 @@
+// Copyright 2021, The Tremor Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{
+    net::{Ipv4Addr, Ipv6Addr},
+    str::FromStr,
+    sync::Arc,
+};
+
+use chrono_tz::Tz;
+pub(super) use clickhouse_rs::types::Value as CValue;
+use either::Either;
+use simd_json::{Value, ValueAccess};
+use tremor_value::Value as TValue;
+use uuid::Uuid;
+
+use super::DummySqlType;
+use crate::errors::{Error, ErrorKind, Result};
+
+const UTC: Tz = Tz::UTC;
+
+pub(super) fn convert_value(
+    column_name: &str,
+    value: &TValue,
+    expected_type: &DummySqlType,
+) -> Result<CValue> {
+    match expected_type {
+        DummySqlType::Array(expected_inner_type) => {
+            // We don't check that all elements of the array have the same type.
+            // Instead, we check that every element can be converted to the expected
+            // array type.
+            wrap_getter_error(column_name, value, ValueAccess::as_array, expected_type)?
+                .iter()
+                .map(|value| convert_value(column_name, value, expected_inner_type))
+                .collect::<Result<Vec<_>>>()
+                .map(|converted_array| {
+                    CValue::Array(
+                        expected_inner_type.as_ref().into(),
+                        Arc::new(converted_array),
+                    )
+                })
+        }
+
+        DummySqlType::Nullable(expected_inner_type) if value.is_null() => {
+            // Null value (when we expect it).
+            Ok(CValue::Nullable(Either::Left(
+                expected_inner_type.as_ref().into(),
+            )))
+        }
+
+        DummySqlType::Nullable(expected_inner_type) => {
+            // TODO: we don't want to `?` here because the expected type it
```

Comment:
i think `?`-ing the error here is fine, isn't it?

Reply:
The "we must return an error if the inner conversion failed" is correct. The only problem is that the error will show something like: `Expected type T, found U`, while it should show `Expected type Nullable(T), found U`.

Reply:
aaaaaah, got it. Makes sense